server/broadcast.go

```diff
@@ -74,24 +80,57 @@ type BroadcastSessionsManager struct {
 func (bsm *BroadcastSessionsManager) selectSession() *BroadcastSession {
 	bsm.sessLock.Lock()
 	defer bsm.sessLock.Unlock()
+	numSess := bsm.sel.Size()
 
 	checkSessions := func(m *BroadcastSessionsManager) bool {
-		numSess := m.sel.Size()
 		if numSess < int(math.Ceil(float64(m.numOrchs)/2.0)) {
 			go m.refreshSessions()
 		}
-		return numSess > 0
+		return (numSess > 0 || bsm.lastSess != nil)
 	}
 	for checkSessions(bsm) {
-		sess := bsm.sel.Select()
-		// Handle the case where there is an error during selection and Select() returns nil when session list length > 0
+		var sess *BroadcastSession
+
+		if bsm.lastSess != nil && len(bsm.lastSess.SegsInFlight) > 0 &&
+			time.Now().Sub(bsm.lastSess.SegsInFlight[0].startTime) < bsm.lastSess.SegsInFlight[0].segDur {
+			// Re-use last session if oldest segment is in-flight for < segDur
+			glog.Warningf("REUSING OLD SESSION")
+			sess = bsm.lastSess
+		} else if numSess > 0 {
+			// Or try a new session from the available ones
+			sess = bsm.sel.Select()
+		}
+
+		// If no new sessions are available, re-use last session when oldest segment is in-flight for < 2 * segDur
+		if sess == nil && bsm.lastSess != nil && len(bsm.lastSess.SegsInFlight) > 0 &&
+			time.Now().Sub(bsm.lastSess.SegsInFlight[0].startTime) < 2*bsm.lastSess.SegsInFlight[0].segDur {
+			glog.Warningf("NO NEW SESSION FOUND SO REUSING OLD SESSION")
+			sess = bsm.lastSess
+		}
+
+		// No session found, return nil
 		if sess == nil {
+			glog.Warningf("NO SESSION FOUND DROPPING SEG")
+			if bsm.lastSess != nil {
+				bsm.lastSess.SegsInFlight = nil
+				bsm.lastSess = nil
+			}
 			return nil
 		}
 
 		if _, ok := bsm.sessMap[sess.OrchestratorInfo.Transcoder]; ok {
+			if monitor.Enabled && bsm.lastSess != nil && bsm.lastSess.OrchestratorInfo.Transcoder != sess.OrchestratorInfo.Transcoder {
+				monitor.OrchestratorSwapped()
+			}
+			bsm.lastSess = sess
+			glog.Warningf("RETURNING SESSION %v", sess.OrchestratorInfo.Transcoder)
 			return sess
+		} else if bsm.lastSess != nil && sess.OrchestratorInfo.Transcoder == bsm.lastSess.OrchestratorInfo.Transcoder {
```

Comment:
If we're returning in the if block, it's common to skip the else if block in Go i.e.

```golang
if ... {
}

if bsm.lastSess != nil && ... {
}
```